File: src/butterflow/typesystem.py
from typing import List, Dict

# ==========================================
# PHASE 1: Type System & Schema Definition
# ==========================================


class Type:
    def matches(self, other: 'Type') -> bool:
        raise NotImplementedError


class AnyType(Type):
    def matches(self, other): return True
    def __repr__(self): return "Any"


class Atomic(Type):
    def __init__(self, name): self.name = name
    def matches(self, other): return isinstance(
        other, Atomic) and self.name == other.name

    def __repr__(self): return self.name


class Generic(Type):
    def __init__(self, base, inner):
        self.base = base
        self.inner = inner

    def matches(self, other):
        return isinstance(other, Generic) and self.base == other.base and self.inner.matches(other.inner)

    def __repr__(self): return f"{self.base}<{self.inner}>"


class Either(Type):
    def __init__(self, options: List[Type]): self.options = options

    def matches(self, other):
        return any(opt.matches(other) for opt in self.options)

    def __repr__(self): return f"Either({self.options})"


class DictType(Type):
    def __init__(self, fields: Dict[str, Type]): self.fields = fields

    def matches(self, other):
        if not isinstance(other, DictType):
            return False
        # Check if all required fields in self exist in other and match types
        for k, t in self.fields.items():
            if k not in other.fields:
                return False
            if not t.matches(other.fields[k]):
                return False
        return True

    def __repr__(self): return f"Dict({self.fields})"


class TupleType(Type):
    def __init__(self, items: List[Type]): self.items = items

    def matches(self, other):
        if not isinstance(other, TupleType) or len(self.items) != len(other.items):
            return False
        return all(s.matches(o) for s, o in zip(self.items, other.items))

    def __repr__(self): return f"Tuple({self.items})"

# The "Operator" type represents a function signature: Operator<Args, ReturnType>


class Operator(Type):
    def __init__(self, args: Type, ret: Type):
        self.args = args  # Can be DictType or TupleType
        self.ret = ret

    def matches(self, other):
        # An Operator matches another if their return types match
        # (Covariance usually, but strict equality for now)
        if not isinstance(other, Operator):
            return False
        return self.ret.matches(other.ret)

    def __repr__(self): return f"Operator<{self.args}, {self.ret}>"

File: src/typesystem.rs
use std::collections::HashMap;
use std::fmt;

#[derive(Debug, Clone, PartialEq)]
pub enum Type {
    Any,
    Atomic(String),
    Generic(String, Box<Type>),
    Either(Vec<Type>),
    Dict(HashMap<String, Type>),
    Tuple(Vec<Type>),
    Operator(Box<Type>, Box<Type>), // Args, Return
}

impl Type {
    pub fn matches(&self, other: &Type) -> bool {
        match (self, other) {
            (Type::Any, _) => true,
            (_, Type::Any) => true, // Assuming Any matches anything in both directions for now? Or just strictly one way? Python: AnyType.matches(other) = True.
            
            (Type::Atomic(n1), Type::Atomic(n2)) => {
                let matches = n1 == n2;
                if !matches && n1 == "Int" { // Debug specific case
                     println!("DEBUG: Atomic check '{}' == '{}' -> {}", n1, n2, matches);
                     println!("DEBUG: n1 bytes: {:?}, n2 bytes: {:?}", n1.as_bytes(), n2.as_bytes());
                }
                matches
            },
            
            (Type::Generic(b1, i1), Type::Generic(b2, i2)) => {
                b1 == b2 && i1.matches(i2)
            }
            
            (Type::Either(opts), other) => {
                opts.iter().any(|opt| opt.matches(other))
            },
            
            // Note: If 'other' is Either, do we check if 'self' matches *any* of 'other's options?
            // Or does 'other' have to be a specific instance that matches 'self' (pattern)?
            // In Python: DeclType.matches(RhsType).
            // If Decl is Either[A, B], and Rhs is A, Returns True.
            // If Decl is A, and Rhs is Either[A, B], Returns False (usually, unless we narrow).
            // Python impl: Either.matches(other) = any(opt.matches(other) ... )
            
            (Type::Dict(f1), Type::Dict(f2)) => {
                // Check if all required fields in self exist in other and match
                for (k, t) in f1 {
                    match f2.get(k) {
                        Some(other_t) => {
                            if !t.matches(other_t) {
                                return false;
                            }
                        }
                        None => return false,
                    }
                }
                true
            }
            
            (Type::Tuple(i1), Type::Tuple(i2)) => {
                if i1.len() != i2.len() {
                    return false;
                }
                i1.iter().zip(i2.iter()).all(|(t1, t2)| t1.matches(t2))
            }
            
            (Type::Operator(a1, r1), Type::Operator(_a2, r2)) => {
                // Return type covariance/match
                r1.matches(r2) 
            }
            
            _ => false,
        }
    }
}

impl fmt::Display for Type {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Type::Any => write!(f, "Any"),
            Type::Atomic(n) => write!(f, "{}", n),
            Type::Generic(b, i) => write!(f, "{}<{}>", b, i),
            Type::Either(opts) => {
                let s: Vec<String> = opts.iter().map(|t| t.to_string()).collect();
                write!(f, "Either([{}])", s.join(", "))
            },
            Type::Dict(fields) => {
                let s: Vec<String> = fields.iter().map(|(k,v)| format!("{}: {}", k, v)).collect();
                write!(f, "Dict({{{}}})", s.join(", "))
            },
            Type::Tuple(items) => {
                let s: Vec<String> = items.iter().map(|t| t.to_string()).collect();
                write!(f, "Tuple(({}))", s.join(", "))
            },
            Type::Operator(a, r) => write!(f, "Operator<{}, {}>", a, r),
        }
    }
}
